### ISSUE TITLE
are foreign app run in individual VM's?

### ISSUE THREAD

**Issue by lonnietc** [phip1611/diplomarbeit-impl#7]
Hello,

I have been reading your thesis and have compiled the code.

Are each of the foreign "hello_world" examples running in their own VM's when loaded?

What I need is to try and see some way to run each of them, or some other one, in a few different VM's as well as to have a pager.

I know that the thesis says that currently there are no screen writes being done, but can you please help how how that might be implemented?

In the past I have been able to compile and run the NOVA NRE and NUL systems from scratch, but want to build similar approaches using your implementation.

Any help would be greatly appreciated, my friend.
Thanks and have a great day

**Comment by phip1611** [comment]
Hey Lonnie,

first, I'd like to inform you about the upstream Hedron changes: https://github.com/cyberus-technology/hedron/blob/master/CHANGELOG.md, as you might not be familiar with them yet.

Since API Version 12, Hedron doesn't handle interrupts anymore and gives the privileged and trusted control VM nearly full access to the hardware. This has the implication that global ECs ("threads") do no longer have preemptive scheduling. Instead, a runtime environment needs cooperative scheduling via the new `HC_EC_YIELD` syscall.

While this itself is not a major hurdle, having no interrupts prevent having a fully-fledged runtime-environment with a disk driver, HID drivers, etc. is.

TL;DR: The test of time showed that Hedron doesn't need a sophisticated runtime environment. Instead, the decision was made to simplify the stack and focus on performant virtualization with a privileged control VM.

---

Let's pretend that global ECs still work like they used to be, and we build a sophisticated and feature-rich runtime environment for Hedron; let's call it X: 

X needs to be able to start VMs. For that, it needs a VMM. You either reuse one that runs under Linux and provide a feature-rich Linux OS personality (see my thesis) or you write a Hedron-native VMM. Both approaches for itself require a big amount of engineering work, and you need experience in the specific field. Especially debugging can be hard.

To see how a VMM can be structured in general, you can look at [Cloud Hypervisor](https://github.com/cloud-hypervisor/cloud-hypervisor/tree/main/hypervisor). But again, to enable that all as part of X, a great amount of engineering work needs to be done.


TL;DR: Even with years of experience in this field, having a modern VMM written in Rust running under Hedron needs a huge amount of designing, planning, and engineering. One does not simply write a feature-rich and sophisticated VMM.

---

It's also not possible to just apply one or two patches to NRE or NUL to enable them under the latest Hedron, as the APIs diverged too much. Sorry if I have to disappoint you here.


I hope this answers your questions. It's a tough domain and not easy* - as much engineering is required.
